### PR TITLE
[优化]跑分17000+ 新增AI模块，机器人自主学习生成跳跃公式

### PR DESCRIPTION
**Pull request by BeiTown**
wx20180103-140426 2x

机器人精确采集跳跃结果并自主学习，使用线性回归方法

拟合出最优 [按压时间]->[弹跳距离] 线性公式 Y =kX + b

本优化无需修改config文件，可以适配所有手机，经过十次以上跳跃学习，机器人即可

模拟出相对稳定的线性公式。随着采集结果越多，跳跃也越精确，后期基本连续命中靶心。

理论上只要目标点获取无误，会一直跳下去

修改后最高分数：17000+ （死于飘音符bug。。。）

wx20180103-173944 2x

### CODE REVIEW

**Comment by Sniper970119** [comment]
ai  345行 不支持此设备是为什么啊

**Comment by F-loat** [comment]
@Sniper970119 获取截图的问题

**Comment by Sniper970119** [comment]
default

**Comment by leniy** [comment]
我把debug函数的新变量放在最后面了。增加兼容性：
以前的代码，没有新参数的，默认赋值为auto
新的ai代码，用的d和t

@BeiTown 变量不要用'type'，这个是python保留字

**Comment by F-loat** [comment]
@Sniper970119 新开issue 吧，这部分你的评论会删除，不然比较乱了

File: wechat_jump_auto_ai.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+under_game_score_y = config['under_game_score_y']
+press_coefficient = config['press_coefficient']       # 长按的时间系数，请自己根据实际情况调节
+piece_base_height_1_2 = config['piece_base_height_1_2']   # 二分之一的棋子底座高度，可能要调节
+piece_body_width = config['piece_body_width']             # 棋子的宽度，比截图中量到的稍微大一点比较安全，可能要调节
+
+
+screenshot_way = 2
+
+
+def pull_screenshot():
```
Comment (F-loat) [inline wechat_jump_auto_ai.py]: 这部分函数建议抽离出来共用吧，目前 `wechat_jump_auto.py` 这个版本里的兼容性应该是最好的，替换换行符不需要判断平台，adb 命令的锅

File: wechat_jump_auto_ai.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+# 核心：每次落稳之后截图，根据截图算出棋子的坐标和下一个块顶面的中点坐标，
+#      根据两个点的距离乘以一个时间系数获得长按的时间
+# 识别棋子：靠棋子的颜色来识别位置，通过截图发现最下面一行大概是一条直线，就从上往下一行一行遍历，
+#      比较颜色（颜色用了一个区间来比较）找到最下面的那一行的所有点，然后求个中点，
+#      求好之后再让 Y 轴坐标减小棋子底盘的一半高度从而得到中心点的坐标
+# 识别棋盘：靠底色和方块的色差来做，从分数之下的位置开始，一行一行扫描，由于圆形的块最顶上是一条线，
+#      方形的上面大概是一个点，所以就用类似识别棋子的做法多识别了几个点求中点，
+#      这时候得到了块中点的 X 轴坐标，这时候假设现在棋子在当前块的中心，
+#      根据一个通过截图获取的固定的角度来推出中点的 Y 坐标
+# 最后：根据两点的坐标算距离乘以系数来获取长按时间（似乎可以直接用 X 轴距离）
```
Comment (binderclip) [inline wechat_jump_auto_ai.py]: 这里求加上 AI 版本的思路，和 auto 版本相同的可以一笔带过

File: wechat_jump_auto_ai.py (hunk 1 above)
Comment (leniy) [reply wechat_jump_auto_ai.py]: 对，这部分的代码，我今天抽离出来放到common的screenshot下面了，直接import引用就行。

File: wechat_jump_auto_ai.py [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+import pandas
+from PIL import Image
+import random
+from six.moves import input
+# try:
+from common import debug, config, ai
+
+# except ImportError:
+#    print('请在项目根目录中运行脚本')
+#   exit(-1)
```
Comment (binderclip) [inline wechat_jump_auto_ai.py]: 这里请取消掉注释，然后按找 auto 里面的更新一下吧

**Comment by HJH12315** [comment]
@BeiTown  ios可以用嘛？

File: wechat_jump_auto_ai.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+
+
+VERSION = "1.1.1"
+
+
+debug_switch = True    # debug 开关，需要调试的时候请改为：True
+config = config.open_accordant_config()
```
Comment (binderclip) [inline wechat_jump_auto_ai.py]: 这个实际没用 config 了，就别从 config 读了，万一有的人的机型没有 config 可能会报错，直接写几个默认的参数如何？

File: wechat_jump_auto_ai.py [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+                piece_y_max = max(i, piece_y_max)
+
+    if not all((piece_x_sum, piece_x_c)):
+        return 0, 0,
+    piece_x = int(piece_x_sum / piece_x_c)
+    piece_y = piece_y_max - piece_base_height_1_2  # 上移棋子底盘高度的一半
+
+    return piece_x,piece_y
+
+def find_piece_and_board(im):
```
Comment (binderclip) [inline wechat_jump_auto_ai.py]: 这些尽量也抽出去，和 auto 公用一份吧

File: wechat_jump_auto_ai.py [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+    pull_screenshot()
+    try:
+        Image.open('./autojump.png').load()
+        print('采用方式 {} 获取截图'.format(screenshot_way))
+    except Exception:
+        screenshot_way -= 1
+        check_screenshot()
+
+
+def yes_or_no(prompt, true_value='y', false_value='n', default=True):
```
Comment (binderclip) [inline wechat_jump_auto_ai.py]: 这个也可以抽出去了

File: wechat_jump_auto_ai.py [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+        elif i == false_value:
+            return False
+        prompt = 'Please input %s or %s: ' % (true_value, false_value)
+        i = input(prompt)
+
+def main():
+    '''
+    主函数
+    '''
+    # op = yes_or_no('请确保手机打开了 ADB 并连接了电脑，然后打开跳一跳并【开始游戏】后再用本程序，确定开始？')
```
Comment (binderclip) [inline wechat_jump_auto_ai.py]: 取消了注释吧

**Comment by binderclip** [comment]
引入 AI 版本看起来更重要些，先能用就行了。
要不抽代码啥的之后再做？
@F-loat @leniy

**Comment by leniy** [comment]
@binderclip 可以，先调AI。

我等下吃完饭回来，估计就赶不上你们的商量进度了，就默默的抽代码吧。我测试完毕后直接提交到BeiTown 的仓库，你们就能实时看到了

**Comment by binderclip** [comment]
实际测的时候不是很好用，不知道游戏结束的时候的脏数据会不会被计算进去。

**Comment by ZedeX** [comment]
实际测了一下，貌似不太行，普遍跳的偏近，而且跳了100就挂了。

另外，【# 在跳跃落下的瞬间 摄像机移动前截图 这个参数要自己校调】
time.sleep(0.2)
这个是干啥的

image
image
image
image
image
image
image
image





Distance | Press_time
-- | --
[573.4] | [764.0]
[592.68] | [810.0]
[514.79] | [689.0]
[346.91] | [490.0]
[427.29] | [581.0]
[546.71] | [731.0]
[541.88] | [716.0]
[586.42] | [799.0]
[560.64] | [777.0]
[342.99] | [438.0]
[572.82] | [786.0]
[511.0] | [675.0]
[597.96] | [840.0]
[420.72] | [529.0]
[294.8] | [350.0]
[384.89] | [548.0]
[538.75] | [736.0]
[476.32] | [614.0]
[533.1] | [735.0]
[547.37] | [738.0]
[286.48] | [403.0]
[439.01] | [518.0]
[453.29] | [681.0]
[632.91] | [861.0]
[578.33] | [808.0]
[590.78] | [836.0]
[285.8] | [385.0]
[679.55] | [939.0]
[577.75] | [811.0]
[674.52] | [922.0]
[540.99] | [747.0]
[407.41] | [542.0]
[438.91] | [597.0]
[439.43] | [539.0]
[609.46] | [838.0]
[641.56] | [893.0]
[402.85] | [560.0]
[665.89] | [899.0]
[530.95] | [722.0]
[266.62] | [340.0]
[464.5] | [596.0]
[311.48] | [406.0]
[450.4] | [631.0]
[642.59] | [911.0]
[448.96] | [602.0]
[635.49] | [856.0]
[581.77] | [845.0]
[375.85] | [482.0]
[401.88] | [545.0]
[107.88] | [241.0]
[506.05] | [703.0]
[398.48] | [575.0]
[410.83] | [512.0]
[338.23] | [466.0]
[474.86] | [645.0]
[261.84] | [326.0]
[419.11] | [497.0]
[540.92] | [716.0]
[570.02] | [788.0]
[360.73] | [364.0]
[483.4] | [617.0]
[409.84] | [550.0]
[347.42] | [410.0]
[415.9] | [518.0]
[545.22] | [772.0]
[606.89] | [839.0]
[396.79] | [536.0]
[430.34] | [555.0]
[446.43] | [593.0]
[380.23] | [512.0]
[638.01] | [869.0]
[483.02] | [608.0]
[653.56] | [881.0]
[295.22] | [351.0]
[288.58] | [348.0]
[417.67] | [561.0]
[618.07] | [860.0]
[431.36] | [569.0]
[260.7] | [373.0]
[367.49] | [531.0]
[89.39] | [252.0]
[529.2] | [698.0]
[428.83] | [560.0]
[469.05] | [597.0]
[329.13] | [426.0]
[454.81] | [588.0]
[562.13] | [769.0]
[329.22] | [369.0]
[382.11] | [453.0]
[329.12] | [411.0]
[622.14] | [838.0]
[320.57] | [413.0]
[579.06] | [755.0]
[594.48] | [850.0]
[608.88] | [837.0]
[311.31] | [417.0]
[533.92] | [726.0]
[405.01] | [465.0]
[383.08] | [467.0]
[452.72] | [566.0]
[442.65] | [583.0]
[538.11] | [760.0]
[408.61] | [541.0]
[345.39] | [412.0]
[597.96] | [801.0]
[416.29] | [538.0]
[467.03] | [563.0]
[668.91] | [922.0]
[541.93] | [745.0]
[445.47] | [586.0]
[547.86] | [693.0]
[593.72] | [826.0]
[378.09] | [509.0]
[610.98] | [843.0]
[596.17] | [825.0]
[531.03] | [735.0]
[551.78] | [698.0]
[416.47] | [565.0]
[479.27] | [683.0]
[328.71] | [444.0]
[422.65] | [572.0]
[431.89] | [573.0]
[417.47] | [621.0]
[531.09] | [696.0]
[483.37] | [632.0]
[573.33] | [779.0]
[393.33] | [486.0]
[197.9] | [333.0]
[456.7] | [617.0]
[518.87] | [670.0]
[280.63] | [335.0]
[309.08] | [365.0]
[646.36] | [892.0]
[462.6] | [634.0]
[474.31] | [620.0]
[585.53] | [853.0]
[401.3] | [433]
[597.87] | [830]
[315.64] | [359]

**Comment by BeiTown** [comment]
需要修改sleep(s)的值，以使得截图时棋子正好落地，而摄像机还未移动，这个时候取值才精确

**Comment by ZedeX** [comment]
@BeiTown 可否理解为棋子落地一瞬底座泛起火花那一刻？这个值是取做什么用的？

**Comment by BeiTown** [comment]
@ZedeX 以此算出跳跃距离，但是可能会有偏差，毕竟截图可能会延迟，摄像机已经开始移动

**Comment by toddlerya** [comment]
time.sleep(0.2) 这个参数不好调，不知道是不是我手机原因，老安卓，魅族mx3，拟合结果很离谱。。。

**Comment by dp199313** [comment]
试了一下ai,手机是小米mix感觉,参数收敛有点慢啊,两个小时训练过去了最高还是二十几分

**Comment by BeiTown** [comment]
@dp199313 你要检查一下debug截图，调整time.sleep(0.2)， 是否正好落地，否则取距离的数据有误差，拟合再多也是没有用的

**Comment by BeiTown** [comment]
@toddlerya 当前靠落地瞬间取距离的方法不是很精确，期待有精确算出真实跳跃距离的方案

**Comment by neciszhang** [comment]
iphone7 plus不支持吗

**Comment by kajweb** [comment]
程序版本号：1.1.1
**********
Screen: Physical size: 720x1280
Density: Physical density: 320
DeviceType: Redmi 4A
OS: darwin
Python: 2.7.10 (default, Feb  7 2017, 00:08:15) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)]
**********
暂不支持当前设备

**Comment by toddlerya** [comment]
进行了一点点优化
https://github.com/toddlerya/WechatJumpAI

**Comment by collinsyu** [comment]
time.sleep(0.2 ）   这个弄不准啊，   跳得远要等的时间多， 跳得近要等的时间短， 不准确的话截图计算距离对误差就大，回归函数的数据就不准了，不知道怎么搞，ios 6s
`if distance>470:
            time.sleep(0.02)
        elif distance>450:
            time.sleep(0.015)
        elif distance>440:
            time.sleep(0.014)
        elif distance>430:
            time.sleep(0.012)
        elif distance>420:
            time.sleep(0.011)
        elif distance>415:
            time.sleep(0.011)
        elif distance>410:
            time.sleep(0.010)
        elif distance>395:
            time.sleep(0.008)
        elif distance>390:
            time.sleep(0.007)
        elif distance>380:
            time.sleep(0.006)
        elif distance>370:
            time.sleep(0.005)
        elif distance>350:
            time.sleep(0.004)
        elif distance>320:
            time.sleep(0.003)`


感觉又是一个回归方程了 ，但是  又不知道怎么计算

**Comment by vvvsrx** [comment]
实测Sony Xperia XZ Premium经过设置time.sleep(0.29),跳到了4000+分.太棒了~

**Pull request merged**